staging/src/k8s.io/apiserver/pkg/server/filters/cors.go

```diff
@@ -40,42 +40,49 @@ func WithCORS(handler http.Handler, allowedOriginPatterns []string, allowedMetho
 	allowedOriginPatternsREs := allowedOriginRegexps(allowedOriginPatterns)
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 		origin := req.Header.Get("Origin")
-		if origin != "" {
-			allowed := false
-			for _, re := range allowedOriginPatternsREs {
-				if allowed = re.MatchString(origin); allowed {
-					break
-				}
+		if len(origin) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		if allowed := isAllowed(origin, allowedOriginPatternsREs); allowed {
+			w.Header().Set("Access-Control-Allow-Origin", origin)
+			// Set defaults for methods and headers if nothing was passed
+			if allowedMethods == nil {
+				allowedMethods = []string{"POST", "GET", "OPTIONS", "PUT", "DELETE", "PATCH"}
+			}
+			if allowedHeaders == nil {
+				allowedHeaders = []string{"Content-Type", "Content-Length", "Accept-Encoding", "X-CSRF-Token", "Authorization", "X-Requested-With", "If-Modified-Since"}
 			}
-			if allowed {
-				w.Header().Set("Access-Control-Allow-Origin", origin)
-				// Set defaults for methods and headers if nothing was passed
-				if allowedMethods == nil {
-					allowedMethods = []string{"POST", "GET", "OPTIONS", "PUT", "DELETE", "PATCH"}
-				}
-				if allowedHeaders == nil {
-					allowedHeaders = []string{"Content-Type", "Content-Length", "Accept-Encoding", "X-CSRF-Token", "Authorization", "X-Requested-With", "If-Modified-Since"}
-				}
-				if exposedHeaders == nil {
-					exposedHeaders = []string{"Date"}
-				}
-				w.Header().Set("Access-Control-Allow-Methods", strings.Join(allowedMethods, ", "))
-				w.Header().Set("Access-Control-Allow-Headers", strings.Join(allowedHeaders, ", "))
-				w.Header().Set("Access-Control-Expose-Headers", strings.Join(exposedHeaders, ", "))
-				w.Header().Set("Access-Control-Allow-Credentials", allowCredentials)
+			if exposedHeaders == nil {
+				exposedHeaders = []string{"Date"}
+			}
+			w.Header().Set("Access-Control-Allow-Methods", strings.Join(allowedMethods, ", "))
+			w.Header().Set("Access-Control-Allow-Headers", strings.Join(allowedHeaders, ", "))
+			w.Header().Set("Access-Control-Expose-Headers", strings.Join(exposedHeaders, ", "))
+			w.Header().Set("Access-Control-Allow-Credentials", allowCredentials)
 
-				// Stop here if its a preflight OPTIONS request
-				if req.Method == "OPTIONS" {
-					w.WriteHeader(http.StatusNoContent)
-					return
-				}
+			// Stop here if its a preflight OPTIONS request
+			if req.Method == "OPTIONS" {
+				w.WriteHeader(http.StatusNoContent)
+				return
 			}
 		}
+
 		// Dispatch to the next handler
 		handler.ServeHTTP(w, req)
 	})
 }
 
+func isAllowed(origin string, allowedOriginPatternsREs []*regexp.Regexp) bool {
+	for _, re := range allowedOriginPatternsREs {
+		if leftRightIndex := re.FindStringIndex(origin); len(leftRightIndex) == 2 && leftRightIndex[1] == len(origin) {
```

Comment:
I'm pretty sure this does not do the right thing when the origin header contains the scheme and/or the port

The :port portion is optional (according to the [RFC](https://www.rfc-editor.org/rfc/rfc6454#page-13))

This reminds me a lot of https://github.com/openshift/origin/pull/16204

Reply:
Thanks @liggitt for the pointer

I made some changes assuming the following:
- we want to proactively prevent this CVE even though the regex defined by the cluster operator does not fully match the host name in origin.
- the kube-apiserver can't change the cluster operator defined regex on the fly, there is no way for us to verify the validity of the changed regex. is it as simple as appending a `(:|$)` to the regexp string?


Alternatively we could:
- No changes in CORS implementation, we can ask the cluster operator to fix their regex to match the entirety of the origin header (how about the case when the origin header has multiple origins?)
- Change the CORS filter to check if the regex matches the entirety of the origin header and report to log or metric that there could be a potential CVE there

Reply:
> we want to proactively prevent this CVE even though the regex defined by the cluster operator does not fully match the host name in origin.

There's multiple categories I can see existing clusters being in:

Unsafe options, for example:
1. literal domain name not realizing it is a regex (e.g. `foo.bar.com`, which can match `fooxbar.com`)
2. regex for a single origin not realizing they are doing a partial match (e.g. `foo\.com`, which can match `foo.companybar.com`)
3. regex for multiple origins not realizing they are doing a partial match (e.g. `(foo\.com|bar\.com)`, which can match `foo.companybar.com`)
4. regex intending to do partial matches not realizing they are also doing partial suffix matches (e.g. `.*\.foo\.com`, which can match `x.foo.companybar.com`)

Safe options, for example:
1. regex which forces matching the whole host (e.g. `//foo\.com(:|$)`)
2. regex which forces matching the whole origin and scheme (e.g. `https://foo\.com(:|$)`)
3. regex which pins to the start/end (e.g. `^https://foo\.com(:|$)`)

We don't want to change anything that would make already safe options stop matching what they already matched. Blindly prepending `//` and appending `(:|$)` to the regex would break those. That only worked in the openshift PR I linked because that was taking literal hostnames as inputs, not regexes of full URLs.

Distinguishing intent when someone has unescaped `.` characters in their config is also tricky. *Probably* it's a bug and they mean't to match a literal `.`, so should have written `/.`, but we can't be 100% sure.

I think trying to auto-fix things in the unsafe category to be safe is going to be really hard to do without any possibility of breaking someone's intended behavior. Would it make more sense to detect when the provided regex doesn't do anything that would pin to the start (missing `^` or `//`) or end (missing `:` or `$`) and fail validation of the flag with a message indicating how they can modify the flag value? That would have the benefit of prompting the operator to make their config safe in a way they would also probably want to apply to older versions.

Reply:
@liggitt please review the changes when you have some time:
- fail validation if the regular expressions in `cors-allowed-origins` does not pin to start/end
- run the regular expression on each origin from the header (the `origin` header may contain multiple origins)